# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [19]:
import pandas as pd
import seaborn as sns
from seaborn import load_dataset

import matplotlib.pyplot as plt

import numpy as np

import statsmodels.formula.api as smf

import numpy as np

In [6]:
tips = sns.load_dataset('tips')

In [12]:
tips['tip_pct'] = tips['tip'] / tips['total_bill']
tips['net_bill'] = tips['total_bill'] - tips['tip']
tips.head()

total_bill   tip     sex smoker  day    time  size   tip_pct  net_bill
0       16.99  1.01  Female     No  Sun  Dinner     2  0.059447     15.98
1       10.34  1.66    Male     No  Sun  Dinner     3  0.160542      8.68
2       21.01  3.50    Male     No  Sun  Dinner     3  0.166587     17.51
3       23.68  3.31    Male     No  Sun  Dinner     2  0.139780     20.37
4       24.59  3.61  Female     No  Sun  Dinner     4  0.146808     20.98

In [27]:
reg = smf.ols('tip_pct ~ net_bill', data = tips).fit()
reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.189
Model:                            OLS   Adj. R-squared:                  0.185
Method:                 Least Squares   F-statistic:                     56.25
Date:                Thu, 09 Jan 2025   Prob (F-statistic):           1.21e-12
Time:                        17:45:11   Log-Likelihood:                 361.93
No. Observations:                 244   AIC:                            -719.9
Df Residuals:                     242   BIC:                            -712.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.2162      0.008     26.401      0.000       0.200       0.232
net_bill      -0.0033      0.000     -7.500      0.000      -0.004      -0.002
==============================================================================
Omnibus:                      227.090   Durbin-Watson:                   1.990
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             8183.493
Skew:                           3.464   Prob(JB):                         0.00
Kurtosis:                      30.512   Cond. No.                         43.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [41]:
reg = smf.ols('tip_pct ~ np.log(net_bill)', data = tips).fit()
reg.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.289
Model:                            OLS   Adj. R-squared:                  0.286
Method:                 Least Squares   F-statistic:                     98.21
Date:                Thu, 09 Jan 2025   Prob (F-statistic):           1.19e-19
Time:                        17:48:24   Log-Likelihood:                 377.99
No. Observations:                 244   AIC:                            -752.0
Df Residuals:                     242   BIC:                            -745.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.3468      0.019     18.198      0.000       0.309       0.384
np.log(net_bill)    -0.0686      0.007     -9.910      0.000      -0.082      -0.055
==============================================================================
Omnibus:                      169.023   Durbin-Watson:                   2.010
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2946.559
Skew:                           2.432   Prob(JB):                         0.00
Kurtosis:                      19.314   Cond. No.                         17.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [43]:
reg = smf.ols('tip_pct ~ net_bill + np.power(net_bill, 2)', data = tips).fit()
reg.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.240
Model:                            OLS   Adj. R-squared:                  0.233
Method:                 Least Squares   F-statistic:                     37.97
Date:                Thu, 09 Jan 2025   Prob (F-statistic):           4.62e-15
Time:                        17:50:25   Log-Likelihood:                 369.85
No. Observations:                 244   AIC:                            -733.7
Df Residuals:                     241   BIC:                            -723.2
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                 0.2752      0.017     16.504      0.000       0.242       0.308
net_bill                 -0.0101      0.002     -5.796      0.000      -0.014      -0.007
np.power(net_bill, 2)     0.0002   3.95e-05      4.021      0.000     8.1e-05       0.000
==============================================================================
Omnibus:                      194.250   Durbin-Watson:                   2.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4790.145
Skew:                           2.842   Prob(JB):                         0.00
Kurtosis:                      23.949   Cond. No.                     2.42e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.42e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [59]:
reg = smf.ols('np.log(tip_pct) ~ net_bill', data = tips).fit()
reg.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        np.log(tip_pct)   R-squared:                       0.244
Model:                            OLS   Adj. R-squared:                  0.241
Method:                 Least Squares   F-statistic:                     77.95
Date:                Thu, 09 Jan 2025   Prob (F-statistic):           2.18e-16
Time:                        18:04:22   Log-Likelihood:                -55.678
No. Observations:                 244   AIC:                             115.4
Df Residuals:                     242   BIC:                             122.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.5264      0.045    -33.658      0.000      -1.616      -1.437
net_bill      -0.0215      0.002     -8.829      0.000      -0.026      -0.017
==============================================================================
Omnibus:                       11.472   Durbin-Watson:                   1.921
Prob(Omnibus):                  0.003   Jarque-Bera (JB):               23.632
Skew:                          -0.141   Prob(JB):                     7.39e-06
Kurtosis:                       4.498   Cond. No.                         43.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [61]:
tip_pct_pred_log = reg.fittedvalues
tip_pct_pred = np.exp(tip_pct_pred_log)
SST = np.sum((tips['tip_pct'] - tips['tip_pct'].mean())**2)  # Soma total dos quadrados
SSR = np.sum((tips['tip_pct'] - tip_pct_pred)**2)            # Soma dos quadrados dos resíduos

R2_tip_pct = 1 - (SSR / SST)
R2_tip_pct

0.1925745423741939

###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [65]:
reg2 = smf.ols('np.log(tip_pct) ~ np.log(net_bill)', data = tips).fit()
reg2.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        np.log(tip_pct)   R-squared:                       0.280
Model:                            OLS   Adj. R-squared:                  0.277
Method:                 Least Squares   F-statistic:                     94.10
Date:                Thu, 09 Jan 2025   Prob (F-statistic):           5.26e-19
Time:                        18:05:38   Log-Likelihood:                -49.670
No. Observations:                 244   AIC:                             103.3
Df Residuals:                     242   BIC:                             110.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept           -0.8373      0.110     -7.614      0.000      -1.054      -0.621
np.log(net_bill)    -0.3874      0.040     -9.700      0.000      -0.466      -0.309
==============================================================================
Omnibus:                       10.347   Durbin-Watson:                   1.965
Prob(Omnibus):                  0.006   Jarque-Bera (JB):               11.432
Skew:                          -0.397   Prob(JB):                      0.00329
Kurtosis:                       3.702   Cond. No.                         17.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [76]:
tip_pct_pred_log2 = reg2.fittedvalues
tip_pct_pred2 = np.exp(tip_pct_pred_log2)

SST2 = np.sum((tips['tip_pct'] - tips['tip_pct'].mean())**2)  # Soma total dos quadrados
SSR2 = np.sum((tips['tip_pct'] - tip_pct_pred2)**2)            # Soma dos quadrados dos resíduos

R2_tip_pct2 = 1 - (SSR2 / SST2)
R2_tip_pct2

0.3219090541707824

###  6. qual modelo resultou no melhor ajuste?

O ultimo modelo onde foi ajustado no log de do tip_pct em função do log de net_bill, pois obteve o maior coeficiente de determinação com 0.3219